src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/Utils/TypeNameExtensions.cs

```diff
@@ -0,0 +1,82 @@
+        ArrayPool<char>.Shared.Return(assemblyQualifiedName.Array!);
+
+        if (typeName is null || (typeName.AssemblyName is null && !payloadOptions.UndoTruncatedTypeNames))
+        {
+            throw new SerializationException(SR.Format(SR.Serialization_InvalidTypeOrAssemblyName, rawName, libraryRecord.LibraryName));
+        }
+
+        if (typeName.AssemblyName is null && payloadOptions.UndoTruncatedTypeNames)
+        {
+            // Sample invalid input that could lead us here:
+            // TypeName: System.Collections.Generic.List`1[[System.String
+            // LibraryName: 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]
+            // Since the flag is ON, we know it's mangling and we provide missing information.
+            typeName = typeName.WithCoreLibAssemblyName();
+        }
+
+        return typeName;
+    }
+
+    internal static TypeName ParseSystemRecordTypeName(this string rawName, PayloadOptions payloadOptions)
+    {
+        ArraySegment<char> assemblyQualifiedName = GetAssemblyQualifiedName(rawName,
+            CoreLibAssemblyName); // We know it's a System Record, so we set the LibraryName to CoreLib
+
+        TypeName.TryParse(assemblyQualifiedName.AsSpan(), out TypeName? typeName, payloadOptions.TypeNameParseOptions);
+        ArrayPool<char>.Shared.Return(assemblyQualifiedName.Array!);
+
+        return typeName ?? throw new SerializationException(SR.Format(SR.Serialization_InvalidTypeName, rawName));
+    }
+
+    internal static TypeName WithCoreLibAssemblyName(this TypeName systemType)
+        => systemType.WithAssemblyName(CoreLibAssemblyName);
+
+    internal static TypeName WithAssemblyName(this TypeName typeName, string assemblyName)
+    {
+        // For ClassWithMembersAndTypesRecord, the TypeName and LibraryName and provided separately,
+        // and the LibraryName may not be known when parsing TypeName.
+        // For SystemClassWithMembersAndTypesRecord, the LibraryName is not provided, it's always mscorlib.
+        // Ideally, we would just create TypeName with new AssemblyNameInfo.
+        // This will be possible once https://github.com/dotnet/runtime/issues/102263 is done.
+
+        ArraySegment<char> assemblyQualifiedName = GetAssemblyQualifiedName(typeName.FullName, assemblyName);
+        TypeName result = TypeName.Parse(assemblyQualifiedName.AsSpan());
+        ArrayPool<char>.Shared.Return(assemblyQualifiedName.Array!);
+
+        return result;
+    }
+
+    private static ArraySegment<char> GetAssemblyQualifiedName(string typeName, string libraryName)
+    {
+        int length = typeName.Length + 1 + libraryName.Length;
+        char[] rented = ArrayPool<char>.Shared.Rent(length);
+
+        typeName.AsSpan().CopyTo(rented);
+        rented[typeName.Length] = ',';
+        libraryName.AsSpan().CopyTo(rented.AsSpan(typeName.Length + 1));
+
+        return new(rented, 0, length);
```

Comment:
```suggestion
        return new ArraySegment<char>(rented, 0, length);
```

https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md, rule 10:

> 10. We only use var when the type is explicitly named on the right-hand side, typically due to either new or an explicit cast, e.g. var stream = new FileStream(...) not var stream = OpenStandardInput().
>
>    * Similarly, target-typed new() can only be used when the type is explicitly named on the left-hand side, in a variable definition statement or a field definition statement. e.g. FileStream stream = new(...);, but not stream = new(...); (where the type was specified on a previous line).

A `return` statement is not in the list of enumerated acceptable places, because the type name is not already on the line.